### ISSUE TITLE
Coredns 1.8.4 and newer are not able to resolve headless k8s services [EKS 1.21]

### ISSUE THREAD

**Issue by jprecuch** [miekg/coredns#5159]
**What happened**:
Today we've upgraded EKS cluster from 1.20 to 1.21 where we were running 1.8.3 coredns. This worked without issues. 1.21.5 works fine with 1.8.3 coredns. 
However when I change image to 1.8.4 coredns (or higher) our containers are not able to talk to each other and we can see errors like. After I downgrade to 1.8.3 everything works again immidiately
```
time="2022-02-07T11:35:20Z" level=error msg="Error pinging mysqld: dial tcp 127.0.0.1:3306: getsockopt: connection refused" source="mysqld_exporter.go:268"
time="2022-02-07T11:35:32Z" level=error msg="Error pinging mysqld: dial tcp 127.0.0.1:3306: getsockopt: connection refused" source="mysqld_exporter.go:268"
time="2022-02-07T11:35:50Z" level=error msg="Error pinging mysqld: dial tcp 127.0.0.1:3306: getsockopt: connection refused" source="mysqld_exporter.go:268"
time="2022-02-07T11:36:02Z" level=error msg="Error pinging mysqld: dial tcp 127.0.0.1:3306: getsockopt: connection refused" source="mysqld_exporter.go:268"
time="2022-02-07T11:36:20Z" level=error msg="Error pinging mysqld: dial tcp 127.0.0.1:3306: getsockopt: connection refused" source="mysqld_exporter.go:268"
```
There is nothing being reported in coredns logs about being bad
```
kubectl logs coredns-76b57bbd64-hcd2m -n kube-system
.:53
[INFO] plugin/reload: Running configuration MD5 = 47d57903c0f0ba4ee0626a17181e5d94
CoreDNS-1.8.6
linux/amd64, go1.16.5, 13a9191e
```
Any breaking changes in 1.8.4 that could cause this difference? Any troubleshooting guide we can test to make 1.8.4+ work properly? 

**What you expected to happen**:
The update to 1.8.4+ should not break anything compared to 1.8.3

**How to reproduce it (as minimally and precisely as possible)**:
EKS 1.21 + coredns with deployment below
**Anything else we need to know?**:

**Environment**:

- the version of CoreDNS:
1.8.4+
- Corefile:
```
kubectl get configmap -n kube-system coredns -oyaml
apiVersion: v1
data:
  Corefile: |
    .:53 {
        errors
        health
        kubernetes cluster.local in-addr.arpa ip6.arpa {
          pods insecure
          fallthrough in-addr.arpa ip6.arpa
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
```
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
EKS 
- Others:
```
kubectl get deployment coredns -oyaml -n kube-system
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "15"
  creationTimestamp: "2021-06-04T16:29:19Z"
  generation: 15
  labels:
    eks.amazonaws.com/component: coredns
    k8s-app: kube-dns
    kubernetes.io/name: CoreDNS
  name: coredns
  namespace: kube-system
  resourceVersion: "67489120"
  uid: e6083dc5-a57e-4d4a-8210-970361d4c3a3
spec:
  progressDeadlineSeconds: 600
  replicas: 2
  revisionHistoryLimit: 10
  selector:
    matchLabels:
      eks.amazonaws.com/component: coredns
      k8s-app: kube-dns
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 1
    type: RollingUpdate
  template:
    metadata:
      annotations:
        eks.amazonaws.com/compute-type: ec2
      creationTimestamp: null
      labels:
        eks.amazonaws.com/component: coredns
        k8s-app: kube-dns
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: beta.kubernetes.io/os
                operator: In
                values:
                - linux
              - key: beta.kubernetes.io/arch
                operator: In
                values:
                - amd64
                - arm64
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - podAffinityTerm:
              labelSelector:
                matchExpressions:
                - key: k8s-app
                  operator: In
                  values:
                  - kube-dns
              topologyKey: kubernetes.io/hostname
            weight: 100
      containers:
      - args:
        - -conf
        - /etc/coredns/Corefile
        image: 602401143452.dkr.ecr.eu-west-1.amazonaws.com/eks/coredns:v1.8.6-eksbuild.1
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 5
          httpGet:
            path: /health
            port: 8080
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: coredns
        ports:
        - containerPort: 53
          name: dns
          protocol: UDP
        - containerPort: 53
          name: dns-tcp
          protocol: TCP
        - containerPort: 9153
          name: metrics
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /health
            port: 8080
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            memory: 170Mi
          requests:
            cpu: 100m
            memory: 70Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - all
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/coredns
          name: config-volume
          readOnly: true
        - mountPath: /tmp
          name: tmp
      dnsPolicy: Default
      priorityClassName: system-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: coredns
      serviceAccountName: coredns
      terminationGracePeriodSeconds: 30
      tolerations:
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      - key: CriticalAddonsOnly
        operator: Exists
      volumes:
      - emptyDir: {}
        name: tmp
      - configMap:
          defaultMode: 420
          items:
          - key: Corefile
            path: Corefile
          name: coredns
        name: config-volume
status:
  availableReplicas: 2
  conditions:
  - lastTransitionTime: "2022-02-04T09:42:07Z"
    lastUpdateTime: "2022-02-04T09:42:07Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: "2021-06-04T16:29:25Z"
    lastUpdateTime: "2022-02-07T12:11:29Z"
    message: ReplicaSet "coredns-6b86db5c6c" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 15
  readyReplicas: 2
  replicas: 2
  updatedReplicas: 2
  ```

**Comment by chrisohaver** [comment]
> `dial tcp 127.0.0.1:3306: getsockopt: connection refused`

FWIW, this error message at least does not indicate a dns failure.

Questions:
* Are you changing anything else other than the version of coredns in the cluster when you upgrade to coredns 1.8.4?
* Are the 1.8.4 coredns pods running and ready?
* Do the 1.8.4 coredns pod logs show any errors or warnings?
* Does everything check out in the troubleshooting guide for 1.8.4: https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/

**Comment by jprecuch** [comment]
Thanks @chrisohaver . The error I've posted is the only that is visible. Nothing else in logs of the components that has trouble to start. Any other hints this update could cause?

1. We changed also kube-proxy and aws-node to latest recommended version. However that works fine with 1.8.3 version. Breaks with 1.8.4 version
2. Yes
3. No all logs are pasted in description.
4. Some outputs from debug
```
$ kubectl exec -ti dnsutils -- cat /etc/resolv.conf
nameserver 172.20.0.10
search default.svc.cluster.local svc.cluster.local cluster.local eu-west-1.compute.internal
options ndots:5

Here are logs of the container that has trouble to start (on of them)
$ kubectl logs --namespace=kube-system -l k8s-app=kube-dns -f |grep 10.39.19.19
[INFO] 10.39.19.19:46702 - 26748 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000100117s
[INFO] 10.39.19.19:35141 - 35989 "A IN localhost.sde.svc.cluster.local. udp 49 false 512" NXDOMAIN qr,aa,rd 142 0.000092441s
[INFO] 10.39.19.19:49677 - 57066 "AAAA IN localhost.sde.svc.cluster.local. udp 49 false 512" NXDOMAIN qr,aa,rd 142 0.000158598s
[INFO] 10.39.19.19:54832 - 13206 "A IN localhost.svc.cluster.local. udp 45 false 512" NXDOMAIN qr,aa,rd 138 0.000143994s
[INFO] 10.39.19.19:36944 - 53964 "AAAA IN localhost.svc.cluster.local. udp 45 false 512" NXDOMAIN qr,aa,rd 138 0.000152024s
[INFO] 10.39.19.19:48282 - 63577 "A IN localhost.cluster.local. udp 41 false 512" NXDOMAIN qr,aa,rd 134 0.000106661s
[INFO] 10.39.19.19:55609 - 39796 "AAAA IN localhost.cluster.local. udp 41 false 512" NXDOMAIN qr,aa,rd 134 0.000112654s
[INFO] 10.39.19.19:39354 - 62725 "A IN localhost.eu-west-1.compute.internal. udp 54 false 512" NXDOMAIN qr,rd,ra 54 0.000351134s
[INFO] 10.39.19.19:45271 - 45434 "AAAA IN localhost.eu-west-1.compute.internal. udp 54 false 512" NXDOMAIN qr,aa,rd,ra 167 0.00005618s
[INFO] 10.39.19.19:57296 - 53814 "A IN localhost. udp 27 false 512" NOERROR qr,aa,rd,ra 52 0.000051343s
[INFO] 10.39.19.19:60672 - 64099 "AAAA IN localhost. udp 27 false 512" NOERROR qr,rd,ra 64 0.000349331s
[INFO] 10.39.19.19:48935 - 3454 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000100921s
[INFO] 10.39.19.19:34533 - 39386 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000254759s
[INFO] 10.39.19.19:54467 - 46625 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000185312s
[INFO] 10.39.19.19:48567 - 49975 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000114s
[INFO] 10.39.19.19:55770 - 42910 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000179533s
[INFO] 10.39.19.19:52172 - 17666 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000114019s
[INFO] 10.39.19.19:53276 - 44619 "A IN localhost.sde.svc.cluster.local. udp 49 false 512" NXDOMAIN qr,aa,rd 142 0.000157775s
[INFO] 10.39.19.19:33034 - 17929 "AAAA IN localhost.sde.svc.cluster.local. udp 49 false 512" NXDOMAIN qr,aa,rd 142 0.000184828s
[INFO] 10.39.19.19:37797 - 27277 "AAAA IN localhost.cluster.local. udp 41 false 512" NXDOMAIN qr,aa,rd 134 0.000076853s
[INFO] 10.39.19.19:56959 - 20224 "A IN localhost.cluster.local. udp 41 false 512" NXDOMAIN qr,aa,rd 134 0.000145131s
[INFO] 10.39.19.19:33972 - 53980 "A IN localhost.svc.cluster.local. udp 45 false 512" NXDOMAIN qr,aa,rd 138 0.000125178s
[INFO] 10.39.19.19:58664 - 17039 "AAAA IN localhost.svc.cluster.local. udp 45 false 512" NXDOMAIN qr,aa,rd 138 0.000115144s
[INFO] 10.39.19.19:44064 - 49620 "A IN localhost.eu-west-1.compute.internal. udp 54 false 512" NXDOMAIN qr,rd,ra 54 0.000422338s
[INFO] 10.39.19.19:50694 - 37489 "AAAA IN localhost.eu-west-1.compute.internal. udp 54 false 512" NXDOMAIN qr,aa,rd,ra 167 0.000057906s
[INFO] 10.39.19.19:54273 - 57806 "AAAA IN localhost. udp 27 false 512" NOERROR qr,aa,rd,ra 64 0.000054783s
[INFO] 10.39.19.19:46931 - 18206 "A IN localhost. udp 27 false 512" NOERROR qr,aa,rd,ra 52 0.000050474s
[INFO] 10.39.19.19:33288 - 7147 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000097707s
[INFO] 10.39.19.19:51150 - 51973 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000182643s
[INFO] 10.39.19.19:55416 - 12208 "SRV IN mysql.sde.svc.cluster.local. udp 45 false 512" NOERROR qr,aa,rd 475 0.000107188s

$ kubectl exec -i -t dnsutils -- nslookup mysql.sde
Server:		172.20.0.10
Address:	172.20.0.10#53

Name:	mysql.sde.svc.cluster.local
Address: 10.39.19.208
Name:	mysql.sde.svc.cluster.local
Address: 10.39.17.85


$ kubectl get pods --namespace=kube-system -l k8s-app=kube-dns
NAME                       READY   STATUS    RESTARTS   AGE
coredns-7cc879f8db-dmf54   1/1     Running   0          11m
coredns-7cc879f8db-gm2gx   1/1     Running   0          11m


$ kubectl get svc --namespace=kube-system

NAME                               TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)         AGE
kube-controller-manager-exporter   ClusterIP   172.20.203.212   <none>        10252/TCP       140d
kube-dns                           ClusterIP   172.20.0.10      <none>        53/UDP,53/TCP   247d
kube-dns-exporter                  ClusterIP   172.20.1.2       <none>        10055/TCP       140d
kube-scheduler-exporter            ClusterIP   172.20.21.7      <none>        10251/TCP       140d

$ kubectl get endpoints kube-dns --namespace=kube-system -owide
NAME       ENDPOINTS                                                     AGE
kube-dns   10.39.17.118:53,10.39.18.134:53,10.39.17.118:53 + 1 more...   247d
```

**Comment by chrisohaver** [comment]
Hmm, nothing looks wrong there.  Looks like CoreDNS is functioning correctly.

Unrelated: I do see a bunch of `localhost` queries reaching CoreDNS, presumably from pods that don't have `localhost` in their hosts file (or ignore their hosts file).  The _local_ plugin might help there.

But other than that, things look perfectly fine for CoreDNS functionally.  It's up and running, and responding correctly to DNS queries.

> Here are logs of the container that has trouble to start

What do you mean by "trouble to start".  Are the CoreDNS pods having trouble starting?

**Comment by jprecuch** [comment]
@chrisohaver We've dig deeper today and found the root cause of this issue (we hope)
It seems starting 1.8.4+ coredns the headless kubernetes services are not being registered by CoreDNS.
Is this known issue or easy to fix?

1.8.4+ does not resolve headless services
```
root@mysql-0:/var/lib/mysql# host mysql
Host mysql not found: 3(NXDOMAIN)
root@mysql-0:/var/lib/mysql# host mongodb
Host mongodb not found: 3(NXDOMAIN)
root@mysql-0:/var/lib/mysql# host auth
auth.default.svc.cluster.local is an alias for heimdall.default.svc.cluster.local.
heimdall.default.svc.cluster.local has address 172.20.149.229
root@mysql-0:/var/lib/mysql# host pgbouncer
pgbouncer.default.svc.cluster.local has address 172.20.147.32
root@mysql-0:/var/lib/mysql# host heimdall
heimdall.default.svc.cluster.local has address 172.20.149.229
```

1.8.3 works fine:
```
root@mysql-0:/var/lib/mysql# host mysql
mysql.default.svc.cluster.local has address 10.0.101.90
root@mysql-0:/var/lib/mysql# host mongodb
mongodb.default.svc.cluster.local has address 10.0.100.11
root@mysql-0:/var/lib/mysql# host auth
auth.default.svc.cluster.local is an alias for heimdall.default.svc.cluster.local.
root@mysql-0:/var/lib/mysql# host heimdall
heimdall.default.svc.cluster.local has address 172.20.210.243

```

Services we use
```
mysql                            ClusterIP      None             <none>                           3306/TCP,9104/TCP                           242d
mongodb                          ClusterIP      None             <none>                           27017/TCP                                   242d
heimdall                         ClusterIP      172.20.77.184    <none>                           80/TCP,6060/TCP,9000/TCP                    242d
auth                             ExternalName   <none>           heimdall.sde.svc.cluster.local   <none>                                      242d
```
The ClusterRole for coreDNS has endpointslices already added
```
- apiGroups:
  - ""
  - discovery.k8s.io
  resources:
  - endpoints
  - services
  - pods
  - namespaces
  - endpointslices
  verbs:
  - list
  - watch
  ```

**Comment by chrisohaver** [comment]
@jprecuch, For the headless services, are the endpoints ready?  There were some releases of coredns that incorrectly created dns records for unready endpoints.  It was fixed around the 1.8.X timeframe.

**Comment by chrisohaver** [comment]
I just verified, this was fixed in 1.8.4. #4580

If the headless services' endpoint addresses are not ready, then this works as designed.

You can verify with `kubectl get endpointslices [slice-name -n namespace] -o yaml`

**Comment by jprecuch** [comment]
@chrisohaver I've dig further into this and it seems that our service were using only deprecated annotation https://github.com/kubernetes/kubernetes/pull/63742
```
service.alpha.kubernetes.io/tolerate-unready-endpoints
```
After I've added to our service 
```
publishNotReadyAddresses: true
```
Everything seems to start fine now. 

Seems this will unblock us using 1.8.4+ coredns. Thanks for your support a lot
